### ISSUE TITLE
@auth with groupsField and groupClaim not working in V2 transformer

### ISSUE THREAD

**Issue by spc16670** [aws-amplify/amplify-category-api#110]

**Comment by hanna-becker** [comment]
> Thanks for this comment, @endoodev1. With this information I was able to create a workaround for our use case until this issue is properly fixed. I added a new "group" property to the Project table that I duplicate the id value into. (This requires the Create query followed by and Update query, as the id is autogenerated, thus unknown at the time of creation.) At least this way we can get all our app's collab features working again and unblock development.
> 
> ```
> type Project
> @model
> @auth(rules: [
>   { allow: owner },
>   { allow: groups, groupsField: "group", operations: [read, update] }
> ])
> {
>   id: ID!
>   group: String! # duplication of id for @auth group claim
>   name: String!
> }
> ```
> 
> Not sure group needs to necessarily be a required field, like I implemented it.

I was in your shoes a couple of weeks ago and luckily found this workaround to be somewhat acceptable for our use case so as to not having to go through the pain of rolling back. Though I wish I had never upgraded to v2 in the first place, because the workarond is obviously less than ideal because of the additional synching overhead and more surface area for partial failures.

I really hope this issue gets fixed, soon!

**Comment by hanna-becker** [comment]
Adding onto this, I noticed that dynamic group auth with custom claims also doesn't work on a field with a custom index. There may already be a separate issue open for this, but I'm adding it here as it looks related. We're working around this in a similar way, by dupicating the project id into a non-indexed field:

```
type ChatMessage
@model
@auth(rules: [
  { allow: owner },
  { allow: groups, groupsField: "group", operations: [read] }
])
{
  id: ID!
  content: String!
  projectID: ID! @index(name: "gsi-Project.chatMessages", sortKeyFields: ["createdAt"], queryField: "chatMessageByProject")
  group: ID! # copy of projectID, as @auth doesn't work on @indexed field
}
```

We would like to see this working again:
```
type ChatMessage
@model
@auth(rules: [
  { allow: owner },
  { allow: groups, groupsField: "projectID", operations: [read] }
])
{
  id: ID!
  content: String!
  projectID: ID! @index(name: "gsi-Project.chatMessages", sortKeyFields: ["createdAt"], queryField: "chatMessageByProject")
}

**Comment by jcary741** [comment]
I am using version 9.2.1 of the Amplify CLI and also have this issue, but was able to workaround it by overriding the auth vtl file for the query (listReportingUnits) that I needed to work.

My schema:
```
type ReportingUnit
@model
@auth(rules: [
    {allow: groups, groups: ["Administrators"], operations: [read]},
    {allow: owner, ownerField: "createdBy", operations: [read]},
    {allow: groups, groupsField: "organizationId", operations: [read]}
]) {
    organizationId: ID! @primaryKey(sortKeyFields: ["reportingUnitId"])
    reportingUnitId: ID!
    label: String
    description: String
    # ...
    createdBy: String
}

type Organization
@model
@auth(rules: [
    {allow: groups, groups: ["Administrators"], operations: [read]},
    {allow: owner, ownerField: "ownedBy", operations: [read] },
    {allow: groups, groupsField: "organizationId", operations: [read]}
]) {
    organizationId: ID! @primaryKey
    label: String
    description: String
    # ...
    ownedBy: String
}
```

`amplify/backend/api/myapi/resolvers/Query.listReportingUnits.auth.1.req.vtl`
```
## [Start] Authorization Steps. **
$util.qr($ctx.stash.put("hasAuth", true))
#set( $isAuthorized = false )
#set( $primaryFieldMap = {} )
#if( $util.authType() == "User Pool Authorization" )
  #if( !$isAuthorized )
    #set( $staticGroupRoles = [{"claim":"cognito:groups","entity":"Administrators"}] )
    #foreach( $groupRole in $staticGroupRoles )
      #set( $groupsInToken = $util.defaultIfNull($ctx.identity.claims.get($groupRole.claim), []) )
      #if( $groupsInToken.contains($groupRole.entity) )
        #set( $isAuthorized = true )
        #break
      #end
    #end
  #end
  #if( !$isAuthorized )
      #set( $groupsInToken = $util.defaultIfNull($ctx.identity.claims.get("cognito:groups"), []) )
      #if( $groupsInToken.contains($ctx.args.organizationId) )
        #set( $isAuthorized = true )
      #else
        #set($errorMessage = "org id was:")
        $util.error("$errorMessage$ctx.args.organizationId")
      #end
  #end
  #if( !$isAuthorized )
    #set( $authFilter = [] )
    #set( $ownerClaim0 = $util.defaultIfNull($ctx.identity.claims.get("sub"), "___xamznone____") )
    #set( $currentClaim1 = $util.defaultIfNull($ctx.identity.claims.get("username"), $util.defaultIfNull($ctx.identity.claims.get("cognito:username"), "___xamznone____")) )
    #set( $ownerClaim0 = "$ownerClaim0::$currentClaim1" )
    #if( $role0 != "___xamznone____" )
      $util.qr($authFilter.add({"createdBy": { "eq": $ownerClaim0 }}))
    #end
    #set( $role0_0 = $util.defaultIfNull($ctx.identity.claims.get("sub"), "___xamznone____") )
    #if( $role0_0 != "___xamznone____" )
      $util.qr($authFilter.add({"createdBy": { "eq": $role0_0 }}))
    #end
    #set( $role0_1 = $util.defaultIfNull($ctx.identity.claims.get("username"), $util.defaultIfNull($ctx.identity.claims.get("cognito:username"), "___xamznone____")) )
    #if( $role0_1 != "___xamznone____" )
      $util.qr($authFilter.add({"createdBy": { "eq": $role0_1 }}))
    #end
    #if( !$authFilter.isEmpty() )
      $util.qr($ctx.stash.put("authFilter", { "or": $authFilter }))
    #end
  #end
#end
#if( !$isAuthorized && $util.isNull($ctx.stash.authFilter) )
$util.unauthorized()
#end
$util.toJson({"version":"2018-05-29","payload":{}})
## [End] Authorization Steps. **
```

Hope this helps!

In case it helps, I've attached the auto-generated vtl file that I'm overriding:
[Query.listReportingUnits.auth.1.req.vtl.txt](https://github.com/aws-amplify/amplify-category-api/files/9365902/Query.listReportingUnits.auth.1.req.vtl.txt)

**Comment by spc16670** [comment]
It has been 8 months since the issue was raised and there is still no official advice from Amplify. The community confirmed the issue and identified bugs in the vtl templates. 

The Amplify docs have not been updated to mention the issue as a known limitation:
https://docs.amplify.aws/cli/graphql/authorization-rules/#how-it-works

Is this going to be fixed? 
Is the official advice never to upgrade and stay on v1? 
Should there be and AWS support ticket raised for this instead of just a Github issue?

**Comment by Alex-Github-Account** [comment]
I am dealing with this bug by self-made walkaroud. The multy-tenancy is critical for serious projects.

**Comment by jbreemhaar** [comment]
@danielleadams Sorry to bump, but any news? It's been months and multiple issues have been created for this problem. Imho this is pretty high prio.

**Comment by dlamon1** [comment]
The answer is to ditch Amplify and start building out a solution that will
offer solutions.

Learning a framework like Sveltekit or Next is not that difficult and will
handle much of the backend / frontend communication.

Mongodb's realm service will handle multi-tenants architecture.

https://www.mongodb.com/community/forums/t/multi-tenant-application-using-node-js-mongo-db/154534

**Comment by kulikala** [comment]
Hi, forks,

I had the same problem with groups strategy.

Generated vtl resolver:
```vtl
  #if( !$isAuthorized )
    #set( $authFilter = [] )
    #foreach( $group in $util.defaultIfNull($ctx.identity.claims.get("Your-groupClaim-field"), []) )
      #if( !$group.isEmpty() )
        $util.qr($authFilter.add({"Your-groupsField-field": { "contains": $group }}))
      #end
    #end
    #if( !$authFilter.isEmpty() )
      $util.qr($ctx.stash.put("authFilter", { "or": $authFilter }))
    #end
  #end
```

I changed the generated Vtl resolver to handle groupClaim with parseJson, 
```vtl
  #if( !$isAuthorized )
    #set( $authFilter = [] )
    #set( $groupClaim0 = $util.defaultIfNull($ctx.identity.claims.get("Your-groupClaim-field"), []) )
    #if( $util.isString($groupClaim0) )
      #if( $util.isList($util.parseJson($groupClaim0)) )
        #set( $groupClaim0 = $util.parseJson($groupClaim0) )
      #else
        #set( $groupClaim0 = [$groupClaim0] )
      #end
    #end
    #foreach( $group in $groupClaim0 )
      #if( !$group.isEmpty() )
        $util.qr($authFilter.add({"Your-groupsField-field": { "contains": $group }}))
      #end
    #end
    #if( !$authFilter.isEmpty() )
      $util.qr($ctx.stash.put("authFilter", { "or": $authFilter }))
    #end
  #end
```
This workaround worked for me.

I guess this code below does the transformation, but I'm not sure...
https://github.com/aws-amplify/amplify-category-api/blob/main/packages/amplify-graphql-auth-transformer/src/resolvers/query.ts#L462

I hope this helps you all.

**Comment by Alex-Github-Account** [comment]
> Hi, forks,
...
> I hope this helps you all.
@kulikala thank you for sharing. But how did you handle the necessity of adding this your VTL code into **every** generated auth vtl?

**Comment by hanna-becker** [comment]
Hey Amplify team, any updates here? It's been almost a year - an awfully long time to wait for a regression fix. Of all the amplify Github issues I'm subscribed to, this has been the most burning one for us. Please let me know if I can help out with additional context, an amplify app id to look into, or whatever is needed.

**Comment by spc16670** [comment]
+1 Amplify team - please give us some official advice. Are we expected to await a fix? Will v1 transformer continue to be supported until the fix arrives?

**Comment by dlamon1** [comment]
The answer to this question is to ditch Amplify.

**Comment by Alex-Github-Account** [comment]
> > Hi, forks,
> > ...
> > I hope this helps you all.
> 
> @kulikala thank you for sharing. But how did you handle the necessity of adding this your VTL code into **every** generated auth vtl?

I handled that by writing a python script which modifies every vtl in a process of every amplify sync. Vtl then restrict access based on whatever tenant id passed within access id token (yes, not the client id). 
But the most appropriate answer is to abandon amplify while you still can.
That is a very hacky and ugly walkaround I am not fan of. But If anyone interested - upvote this to let me know I'd take some time to refactor and share solution on github. At least it makes amplify work for multi-tenant apps.

**Comment by spc16670** [comment]
Can someone reading this and having access to paid AWS support, log a ticket please and escalate to AWS perhaps?

**Comment by naedx** [comment]
I am having this issue again but not in all situations.

With my schema as something like what is shown below (simplified to illustrate the problem), I get a `*.vtl` resolver which parses the claims on `custom:tenant_admin` as JSON but not `custom:subteam_create_*`

```gql

type Task 
    queries: {get: "getTask", list: null},
    subscriptions: { level: off }, 
    timestamps: {createdAt: "createdAt", updatedAt: "updatedAt"}
  @auth(rules: [
    { allow: groups, groupsField: "tid_tenantId", groupClaim: "custom:tenant_admin", operations: [create, get, list, update, delete] }
    
    { allow: groups, groupsField: "tid_subteamGroupIds", groupClaim: "custom:subteam_create", operations: [create] }
    { allow: groups, groupsField: "tid_subteamGroupIds", groupClaim: "custom:subteam_read", operations: [get, list] }
    { allow: groups, groupsField: "tid_subteamGroupIds", groupClaim: "custom:subteam_list", operations: [get, list] }
    { allow: groups, groupsField: "tid_subteamGroupIds", groupClaim: "custom:subteam_update", operations: [update] }
    { allow: groups, groupsField: "tid_subteamGroupIds", groupClaim: "custom:subteam_delete", operations: [delete] }
  ]) 
{
  id: ID!
    @primaryKey(sortKeyFields: ["tenantId"])

  tenantId: ID!
    @index(name: "byTidByStatus", sortKeyFields: ["status"], queryField: "listTasksByTid")

  tid_tenantId: String # duplicate of above field since the key field cannot be used for auth
  
  subteamGroupIds: [String]
  tid_subteamGroupIds: [String] # duplicate of above field since the key field cannot be used for auth

  title: String!
  details: String

  status: String
}


```

```vtl

## [Start] Authorization Steps. **
$util.qr($ctx.stash.put("hasAuth", true))
#set( $isAuthorized = false )
#set( $primaryFieldMap = {} )
#if( $util.authType() == "User Pool Authorization" )
  #if( !$isAuthorized )
    #set( $authFilter = [] )
    #set( $role0 = $util.defaultIfNull($ctx.identity.claims.get("custom:tenant_admin"), []) )
    #if( $util.isString($role0) )
      #if( $util.isList($util.parseJson($role0)) )
        #set( $role0 = $util.parseJson($role0) )
      #else
        #set( $role0 = [$role0] )
      #end
    #end
    #if( !$role0.isEmpty() )
      $util.qr($authFilter.add({ "tid_subteamGroupIds": { "in": $role0 } }))
    #end
    #foreach( $group in $util.defaultIfNull($ctx.identity.claims.get("custom:subteam_create_read"), []) )
      #if( !$group.isEmpty() )
        $util.qr($authFilter.add({"tid_subteamGroupIds": { "contains": $group }}))
      #end
    #end
    #foreach( $group in $util.defaultIfNull($ctx.identity.claims.get("custom:subteam_create_list"), []) )
      #if( !$group.isEmpty() )
        $util.qr($authFilter.add({"tid_subteamGroupIds": { "contains": $group }}))
      #end
    #end
    #foreach( $group in $util.defaultIfNull($ctx.identity.claims.get("custom:subteam_create_readOwn"), []) )
      #if( !$group.isEmpty() )
        $util.qr($authFilter.add({"tid_subteamGroupIds": { "contains": $group }}))
      #end
    #end
    #foreach( $group in $util.defaultIfNull($ctx.identity.claims.get("custom:subteam_create_listOwn"), []) )
      #if( !$group.isEmpty() )
        $util.qr($authFilter.add({"tid_subteamGroupIds": { "contains": $group }}))
      #end
    #end
    #if( !$authFilter.isEmpty() )
      $util.qr($ctx.stash.put("authFilter", { "or": $authFilter }))
    #end
  #end
#end
#if( !$isAuthorized && $util.isNull($ctx.stash.authFilter) )
$util.unauthorized()
#end
$util.toJson({"version":"2018-05-29","payload":{}})
## [End] Authorization Steps. **

```

**Comment by naedx** [comment]
Related to issue #1345

**Comment by mewtlu** [comment]
Couldn't help but notice this issue was re-assigned, any chance we could get an update on if/when this might be actioned? This is still a critical issue in my opinion as it is forcing us to manually maintain a very large number of resolvers.

**Comment by jcary741** [comment]
@mewtlu I agree that maintaining resolvers is horrible and not the experience I wanted from Amplify. This issue alone pushed me to move off of AppSync for my PWA by setting up my own lambda API endpoints for querying data. A few months later, I ditched Amplify entirely for a GitLab CI + terraform build/deploy process. It's more stable, deploys faster, and I have a higher confidence I'm not accidentally leaking extra data because of overcomplicated vtl resolvers.

**Comment by spc16670** [comment]
Same here, we are ditching Amplify because of this issue, we have half of the entities moved to a REST endpoint we expose using a lambda, working on the other half. Big disappointment with Amplify.

**Comment by julien-tamade** [comment]
I'm having this issue too. The resolver for my delete mutation is completely ignoring any cognito auth when i use these field defined groups

```
#if( $util.authType() == "User Pool Authorization" )

#end
```

even though I have defined

```
{allow: owner, ownerField: "userId", identityClaim: "username", operations: [read, create, update, delete]}
  {allow: public, provider: apiKey, operations: [read, update]}
  {allow: public, provider: iam, operations: [read, update, delete]}
  {allow: groups, groups: ["admin"]}
  {allow: groups, groupsField: "groupsCanAccess", operations: [read]}
  {allow: groups, groupsField: "groupsCanModerate", operations: [read, create, update, delete]}])
```

**Comment by Superpat** [comment]
So we got bit by this issue today when we try to introduce dynamic auth in our app. I'm considering migrating to amplify gen 2 earlier than we had planned, but can anyone confirm that this is not an issue in gen 2 ? 

The docs specify that group claims are still a thing: https://docs.amplify.aws/vue/build-a-backend/data/customize-authz/configure-custom-identity-and-group-claim/#pageMain but they dont make it clear if I can use arbitrary information from the model when doing the authorization.

**Comment by coreyroach** [comment]
This may be relevant to some. I ran into the same / similar issue and found that in v6, the `Amplify.configure` setting for `graphql_headers` has been removed. I haven't seen this documented, but just digging through the Typescript definitions found that `Amplify.configure` has a second argument for `libraryOptions`. The interface for `LibraryOptions` includes options for `API.GraphQL.headers`. So, to use the idToken, I updated my configuration as:

```typescript
import { Amplify } from 'aws-amplify';
import { fetchAuthSession } from 'aws-amplify/auth';

//...

Amplify.configure({ ...awsConfig }, {
  API: {
    GraphQL: {
      headers: async () => {
        const session = await fetchAuthSession();
        const idToken = session.tokens.idToken.toString();
        return {
          Authorization: idToken,
        };
      }
    }
  }
});
```
And for what it's worth, my `groupsField` is of type String and not a String set as I've seen in a few examples here.
